File: 백준/Gold/2023. 신기한 소수/신기한 소수.cs
using StreamWriter sw = new(Console.OpenStandardOutput());
int count = int.Parse(Console.ReadLine()), max = GetMax(count);

dfs(0, 0);

void dfs(int depth, int number)
{
    if (depth == count)
    {
        sw.WriteLine(number);
        return;
    }

    for(int i = 1; i <= 9; i++)
    {
        int next = number * 10 + i;
        if(IsPrime(next)) dfs(depth + 1, number * 10 + i);
    }
}

bool IsPrime(int x)
{
    if (x is 1 or < 1) return false;
    if (x is 2) return true;
    if (x % 2 is 0) return false;

    var max = (int)Math.Floor(Math.Sqrt(x));
    
    for(int i = 3; i < x; i += 2)
    {
        if(x % i is 0)
            return false;
    }

    return true;
}

int GetMax(int target)
{
    int acc = 1;
    for(int i = 0; i < target; i++)
    {
        acc = acc * 10;
    }
    return acc;
}

File: 백준/Gold/2023. 신기한 소수/신기한 소수.cc
#include <iostream>
using namespace std;

int n;
void dfs(int, int);
bool isPrime(int);

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n;

    for (auto i : {2, 3, 5, 7})
        dfs(i, 1);

}

void dfs(int num, int depth)
{
    if (depth == n)
    {
        cout << num << '\n';
        return;
    }

    for (int i = 0; i <= 9; i++)
    {
        if (i % 2 == 0)
            continue;
        if (isPrime(num * 10 + i))
            dfs(num * 10 + i, depth + 1);
    }
}

bool isPrime(int num)
{
    for (int i = 2; i <= num / 2; i++)
    {
        if (num % i == 0)
            return false;
    }
    return true;
}